File: typescript/core/src/fast-queue.ts
export class FastQueue {
  private buffer: bigint[];
  private head: number;
  private tail: number;
  private _length: number;
  private _mask: number;

  constructor(initialCapacity = 16) {
    this.buffer = new Array(initialCapacity);
    this.head = 0;
    this.tail = 0;
    this._length = 0;
    this._mask = initialCapacity - 1;
  }

  get length(): number {
    return this._length;
  }

  private expand() {
    const newCapacity = this.buffer.length * 2;
    const newBuffer = new Array(newCapacity);
    
    if (this._length > 0) {
      if (this.head < this.tail) {
        for (let i = 0; i < this._length; i++) {
          newBuffer[i] = this.buffer[this.head + i];
        }
      } else {
        const rightLength = this.buffer.length - this.head;
        for (let i = 0; i < rightLength; i++) {
          newBuffer[i] = this.buffer[this.head + i];
        }
        for (let i = 0; i < this.tail; i++) {
          newBuffer[rightLength + i] = this.buffer[i];
        }
      }
    }

    this.buffer = newBuffer;
    this.head = 0;
    this.tail = this._length;
    this._mask = newCapacity - 1;
  }

  push(...items: bigint[]): void {
    this.pushArray(items);
  }

  pushArray(items: bigint[]): void {
    for (let i = 0; i < items.length; i++) {
      if (this._length === this.buffer.length) {
        this.expand();
      }
      this.buffer[this.tail] = items[i];
      this.tail = (this.tail + 1) & this._mask;
      this._length++;
    }
  }

  unshift(...items: bigint[]): void {
    this.unshiftArray(items);
  }

  unshiftArray(items: bigint[]): void {
    while (this._length + items.length > this.buffer.length) {
      this.expand();
    }
    
    this.head = (this.head - items.length) & this._mask;
    for (let i = 0; i < items.length; i++) {
      this.buffer[(this.head + i) & this._mask] = items[i];
    }
    this._length += items.length;
  }

  shift(): bigint | undefined {
    if (this._length === 0) return undefined;
    const item = this.buffer[this.head];
    this.head = (this.head + 1) & this._mask;
    this._length--;
    return item;
  }

  pop(): bigint | undefined {
    if (this._length === 0) return undefined;
    this.tail = (this.tail - 1) & this._mask;
    const item = this.buffer[this.tail];
    this._length--;
    return item;
  }

  get(index: number): bigint | undefined {
    if (index < 0 || index >= this._length) return undefined;
    return this.buffer[(this.head + index) & this._mask];
  }
}

File: go/src/engine/fast_queue.go
package engine

import (
	. "math/big"
)

type FastQueue struct {
	buffer []*Int
	head   int
	tail   int
	length int
	mask   int
}

func NewFastQueue(initialCapacity int) *FastQueue {
	if initialCapacity < 1 {
		initialCapacity = 16
	}
	capacity := 1
	for capacity < initialCapacity {
		capacity <<= 1
	}
	return &FastQueue{
		buffer: make([]*Int, capacity),
		head:   0,
		tail:   0,
		length: 0,
		mask:   capacity - 1,
	}
}

func (q *FastQueue) expand() {
	newCapacity := len(q.buffer) * 2
	newBuffer := make([]*Int, newCapacity)

	if q.length > 0 {
		if q.head < q.tail {
			for i := 0; i < q.length; i++ {
				newBuffer[i] = q.buffer[q.head+i]
			}
		} else {
			rightLength := len(q.buffer) - q.head
			for i := 0; i < rightLength; i++ {
				newBuffer[i] = q.buffer[q.head+i]
			}
			for i := 0; i < q.tail; i++ {
				newBuffer[rightLength+i] = q.buffer[i]
			}
		}
	}

	q.buffer = newBuffer
	q.head = 0
	q.tail = q.length
	q.mask = newCapacity - 1
}

func (q *FastQueue) Push(item *Int) {
	if q.length == len(q.buffer) {
		q.expand()
	}
	q.buffer[q.tail] = item
	q.tail = (q.tail + 1) & q.mask
	q.length++
}

func (q *FastQueue) UnshiftArray(items []*Int) {
	if len(items) == 0 {
		return
	}
	for q.length+len(items) > len(q.buffer) {
		q.expand()
	}

	q.head = (q.head - len(items)) & q.mask
	for i := 0; i < len(items); i++ {
		q.buffer[(q.head+i)&q.mask] = items[i]
	}
	q.length += len(items)
}

func (q *FastQueue) Shift() *Int {
	if q.length == 0 {
		return nil
	}
	item := q.buffer[q.head]
	q.buffer[q.head] = nil // Avoid leaking reference
	q.head = (q.head + 1) & q.mask
	q.length--
	return item
}

func (q *FastQueue) Pop() *Int {
	if q.length == 0 {
		return nil
	}
	q.tail = (q.tail - 1) & q.mask
	item := q.buffer[q.tail]
	q.buffer[q.tail] = nil // Avoid leaking reference
	q.length--
	return item
}

func (q *FastQueue) Len() int {
	return q.length
}

func (q *FastQueue) Get(index int) *Int {
	if index < 0 || index >= q.length {
		return nil
	}
	return q.buffer[(q.head+index)&q.mask]
}
